app/views.py

```diff
@@ -11,27 +11,56 @@
 from django.views.generic.edit import FormView, UpdateView
 
 from django.contrib.contenttypes.models import ContentType
-from django.shortcuts import get_object_or_404
+from django.shortcuts import get_object_or_404, redirect
 
 
 class IndexView(TemplateView):
     template_name = "explore.html"
 
+    def dispatch(self, request, *args, **kwargs):
+        if request.GET.get("type") not in ["challenge", "project", None]:
+            return redirect("index")
+        return super(IndexView, self).dispatch(request, *args, **kwargs)
+
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        if "type" not in self.request.GET or "q" not in self.request.GET:
-            context["challenges"] = Challenge.objects.all()[:3]
-            context["projects"] = Project.objects.all()[:9]
-            return context
-        query = self.request.GET["q"]
-        if self.request.GET["type"] == "challenge":
-            context["challenges"] = Challenge.objects.all().filter(
-                Q(name__contains=query) | Q(description__contains=query)
-            )
-        elif self.request.GET["type"] == "project":
-            context["projects"] = Project.objects.all().filter(
-                Q(name__contains=query) | Q(description__contains=query)
-            )
+
+        context["tags"] = Tag.objects.all()
+        context["selected_tags"] = []
+
+        if "type" not in self.request.GET or (
+            "q" not in self.request.GET and "tag" not in self.request.GET
+        ):
+            context["objects"] = {
+                "challenge": Challenge.objects.all()[:3],
+                "project": Project.objects.all()[:9],
+            }
```

Comment:
`return` immediately so you can remove the indentation that comes with the `else`.